### PR TITLE
trackpad gesture support

### PR DESCRIPTION
**Pull request by torleif**
This change allows for users who have a trackpad to use it to pan around the battlefield.

I use a macbook that has a trackpad. Using the two fingers to pan makes using OpenRA much more enjoyable. (And would be the same for tablet style devices)

On a side note, I did notice there was a zoom functionality there too. Using pinch to zoom would be pretty cool too.

### CODE REVIEW

**Comment by Punsho** [comment]
Just incase you don't know you can scroll around the map by moving your mouse to the edge of the screen. The default speed is too low for it be the preferred method but you can easily increase it in the settings. 

I'm not opposed to adding more / improving control to other non-mouse&keyboard input devices. In fact I think the game lacks is an inbuilt ability to rebind middle mouse button to something else as touchpads simply can't send that input. The biggest problem with adding these is that it increases maintenance and as most people use a mouse (or if not they have an easy way of obtaining one) is not an urgent one

**Comment by pchote** [comment]
The basic approach here looks reasonable, but there are two additions that IMO are essential before we merge this:
* Pinch-to-zoom: As you've already noticed, we now support smooth zooming and this is going to be one of the headline features of the next release. Adding trackpad support for scrolling while ignoring zooming would feel very out of place.
* Settings for players to customize or disable the gestures: We provide options for players to invert the middle-mouse panning direction, and to select a modifier for mouse-wheel zoom. We should provide similar settings for trackpads too. It may be best to organize the input settings into general settings at the top, then mouse-specific settings under a Mouse heading, then trackpad specific settings under a Trackpad heading.

**Comment by torleif** [comment]
@pchote do you think a tick box to enable or disable gestures in settings would be suffice? I can't think of any other uses gestures could be for at this point, other than panning and zooming

File: OpenRA.Game/Input/IInputHandler.cs [hunk 1]
```diff
@@ -19,9 +19,10 @@ public interface IInputHandler
 		void OnKeyInput(KeyInput input);
 		void OnMouseInput(MouseInput input);
 		void OnTextInput(string text);
+		void OnGestureInput(MouseInput input);
 	}
 
-	public enum MouseInputEvent { Down, Move, Up, Scroll }
+	public enum MouseInputEvent { Down, Move, Up, Scroll, Gesture, FingerDown, FingerUp }
```
Comment (teinarss) [inline OpenRA.Game/Input/IInputHandler.cs]: Would it make sense to split these new values to a `GestureInputEvent` and make a new struct, `GestureInput` similar to `MouseInput` with only the properties needed for this use case?

**Review by teinarss — CHANGES_REQUESTED** [review]

File: OpenRA.Game/Input/IInputHandler.cs (hunk 1 above)
Comment (pchote) [reply OpenRA.Game/Input/IInputHandler.cs]: Sounds reasonable to me.

**Comment by pchote** [comment]
We can start with a checkbox if that makes it easier for you while working on the zoom implementation, but I think we should aim to have the invert-pan toggle and possibly the zoom modifier (depending on how sensitive the zoom gesture triggers) for the final version.

**Comment by pchote** [comment]
I have attempted to address the settings setup with #17586. This PR will just need to make use of the zoom modifier and pan behaviour settings.

**Comment by torleif** [comment]
@pchote thanks for your feed back. I've implemented pinch to zoom, and refactored the code as requested. 

Currently this branch uses RightMouseScroll to detect if panning should be inverted. This will need to be changed to MouseScroll to work with  #17586, depending on what order you want to merge.

**Comment by pchote** [comment]
Sorry for the delay in getting back to this. The panning feels good, but pinching to zoom doesn't seem to work properly - if I pinch slowly it will never change, and if the zoom sensitivity is not set to max the zoom change will stop part way through even though I am continuing to pinch.

I also realized that my suggestion to tie the direction and sensitivity to the settings options is probably not what the player wants if they are using a touch screen. In that case i expect that we want to enforce that the pixels on the map stay fixed to the finger contact points. We would ideally get some feedback from actual touchscreen users when testing this - @cjshmyr do you still have your surface? Does SDL provide a mechanism for distinguishing between a direct touch gesture (i.e. screen) and an indirect one (i.e. touchpad)?

**Comment by pchote** [comment]
I don't have the bandwidth to pick this up in the near future, and I'm not sure that any other regular contributor has the interest. Moving to the Future milestone so we can pick this up again later.

**Pull request closed without merging**